### DISCUSSION TITLE
Mock trait method with Fn(T) + Clone + Send + 'static

### DISCUSSION THREAD

**Discussion by grongor** [asomers/mockall#604] [Questions]
Hello :wave: 

I have a trait that takes a closure which it later uses in multiple places on multiple threads. `mockall` generates argument types like `Fn(Box<dyn Fn(T, bool)>)` so I'm unable to use non-`*_st` methods like `returning`.

Simplified example:
```rust
#[cfg_attr(test, automock)]
pub trait Pinger<T: Send + 'static> {
    fn start<F: Fn(T, bool) + Clone + Send + 'static>(&mut self, done_fn: F);
    // ...other non-relevant methods
}

#[cfg(test)]
mod test {
    use std::sync::{Arc, OnceLock};

    use crate::pinger::MockPinger;

    #[tokio::test]
    async fn test_some_other_thing() {
        let done_fn = Arc::new(OnceLock::new());
        let mut pinger = MockPinger::<bool>::new();
        {
            let done_fn = Arc::clone(&done_fn);
            pinger.expect_start().returning(move |f| assert!(done_fn.set(f).is_ok()));
        }

        // setup other mock methods to call done_fn (this trait is used internally by the tested object)
        
        // let thing = OtherThing::new(pinger);
        // assert!(thing.do_stuff());
    }
}
```

The error I'm encoutering:
```
error[E0277]: `dyn Fn(bool, bool)` cannot be sent between threads safely
   --> crates/lib/src/pinger.rs:218:45
    |
218 |             pinger.expect_start().returning(move |f| assert!(done_fn.set(f).is_ok()));
    |                                   --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `dyn Fn(bool, bool)` cannot be sent between threads safely
    |                                   |
    |                                   required by a bound introduced by this call
    |
    = help: the trait `std::marker::Send` is not implemented for `dyn Fn(bool, bool)`, which is required by `{closure@crates/lib/src/pinger.rs:218:45: 218:53}: std::marker::Send`
    = note: required for `std::ptr::Unique<dyn Fn(bool, bool)>` to implement `std::marker::Send`
note: required because it appears within the type `Box<dyn Fn(bool, bool)>`
   --> /home/grongor/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:237:12
    |
237 | pub struct Box<
    |            ^^^
    = note: required for `OnceLock<Box<dyn Fn(bool, bool)>>` to implement `std::marker::Sync`
    = note: 1 redundant requirement hidden
    = note: required for `Arc<OnceLock<Box<dyn Fn(bool, bool)>>>` to implement `std::marker::Send`
note: required because it's used within this closure
   --> crates/lib/src/pinger.rs:218:45
    |
218 |             pinger.expect_start().returning(move |f| assert!(done_fn.set(f).is_ok()));
    |                                             ^^^^^^^^
note: required by a bound in `__mock_MockPinger_Pinger_13065831140148901594::__start::Expectation::<T>::returning`
   --> crates/lib/src/pinger.rs:201:18
    |
201 | #[cfg_attr(test, automock)]
    |                  ^^^^^^^^ required by this bound in `Expectation::<T>::returning`
```

For now I'm ok with `Rc<OnceCell<_>>` and the `_*st` methods, but I wonder if I'm doing something wrong or if this is some limitation of `mockall` or the `automock` derive attr.

**Comment by asomers** [comment]
So the closure that you're passing to `returning` isn't `Send`.  Why it isn't is a little harder to determine.  What happens if you add the `Sync` bound to T and/or F ?

**Reply by grongor** [reply]
I don't think the closure passed to the `returning` is the issue (and because it is move and only uses `let done_fn = Arc::clone(&done_fn);`, which points to `OnceLock`, I'd say it is `Send`).

Adding `Sync` bound to either has no effect.

**Comment by asomers** [comment]
There are ~~two~~ three problems:

1. You've found a bug.  Mockall isn't handling the additional bounds on `F` correctly.  I'll fix that shortly.
2. The `Clone` trait isn't an automatic trait, so you can't turn `F` into a trait object.  Mockall relies on being able to do that.
3. `F` does also need `Sync`

For a solution, you can either wait for my patch and remove the `Clone`, or you can use `returning_st` instead of `returning`.  It has the constraint that the function must be called from the same thread that set the expectation, but it doesn't require the closure to be `Send`.

**Comment by asomers** [comment]
@grongor try the branch in #606 .  You'll still probably have to remove the `+ Clone`.

**Reply by grongor** [reply]
Thanks for the quick fix! I tried it and it seems to add `Send` properly now. The thing is that I need the `Clone` there :smile: and with the patch it looks like I'm not able to generate the mock at all (`error[E0225]: only auto traits can be used as additional traits in a trait object`), so...what are my options here? Am I able to generate the mock some other way (instead of the `automock` derive) or do I have to skip `mockall` entirely for this specific case?

**Reply by asomers** [reply]
That error is a limitation of the Rust language, not of Mockall.  So I can't fix it for you.  However, I think you can solve this by creating a new trait and then define `Pinger` in terms of that.  Like this:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f412bdb5680b0bb26b0f38a8cd0f9a6
